timelock/program/src/state/enums.rs

```diff
@@ -0,0 +1,82 @@
+/// What kind of consensus algo a timelock uses
+#[derive(Clone, Debug, PartialEq)]
+pub enum ConsensusAlgorithm {
+    /// Run if 51% of tokens are burned in favor of the timelock set
+    Majority,
+    /// Run if 66% of tokens are burned in favor
+    SuperMajority,
+    /// Run only if 100% of tokens are burned in favor
+    FullConsensus,
+}
+
+impl Default for ConsensusAlgorithm {
+    fn default() -> Self {
+        ConsensusAlgorithm::Majority
+    }
+}
+
+/// What type of execution a timelock is
+#[derive(Clone, Debug, PartialEq)]
+pub enum ExecutionType {
+    /// Each transaction is independent of the other
+    Independent,
+}
+
+impl Default for ExecutionType {
+    fn default() -> Self {
+        ExecutionType::Independent
+    }
+}
+
+/// What state a timelock set is in
+#[derive(Clone, Debug, PartialEq)]
+pub enum TimelockStateStatus {
+    /// Draft
+    Draft,
+    /// Taking votes
+    Voting,
+
+    /// Votes complete, in execution phase
+    Executing,
+
+    /// Completed, can be rebooted
+    Completed,
+
+    /// Deleted
+    Deleted,
+
+    /// Defeated
+    Defeated,
+}
+
+impl Default for TimelockStateStatus {
+    fn default() -> Self {
+        TimelockStateStatus::Draft
+    }
+}
+
+/// What type a timelock is
+#[derive(Clone, Debug, PartialEq)]
+pub enum TimelockType {
+    /// Meaning voters deposit their tokens from outside mint and "govern" the proposal
```

Comment:
I think always resorting to rolling a new program might not be the best way to approach this. People will not want to remake their stuff or migrate their stuff. It would be nice to gracefully handle new options without having to go nuclear, no?

Reply:
It depends if you'll allow people to update their existing timelock to be a new type.  Will it make sense to update a timelock from a governance to a "SomethingElse"?

Either way, what other options are you imagining for this enum?  If it's something very different, it may be better served by another program.  Or not!  Mainly trying to understand the rationale for this

Reply:
I think there may be other types of proposals/timelocks besides governance timelocks - perhaps there may be a submission to the council mint to resolve a dispute, which may have an entirely different state machine behind it. Either way, giving up a single byte in the struct for code cleanliness at the expense of future extensibility seems a bit risky.

Reply:
It's the "entirely different state machine" part that seems like it might not fit with this model and be better served with a new program.  Anyway, if you feel strongly about other options for this, it can't hurt.

Reply:
Deploying an entirely new program would be devastating for the governance ecosystem. The migration would be nearly impossible to achieve and require a lot of buy in from too many parties. As a real life example, there are still many swap pools operating on old swap, and to this day the swap UI has to maintain logic for it.

I think we do ourselves a disservice when we assume the solution to every needed change is "deploy a new program". Even if we need to deploy a new state machine in an old program, it's something we can do with some handling - like with this enum. I also was just choosing an arbitrary example here, I don't expect to be building a new state machine any time soon but I'd like to maintain the option!

Reply:
The question isn't whether this program should or shouldn't be upgradable -- I think everyone agrees it should be upgradable!

Swap v1 is a different mess because the program isn't upgradeable at all, and not because we didn't future-proof the code, which is a different consideration.

You're imagining a situation in which we create a new timelock type that isn't governance, and an existing governance timelock wants to transition to using that instead, correct?  I'm not sure how that would work at all, but you may have a clearer image of that.

Either way, agreed that it doesn't hurt and that it's just 1 byte.

Reply:
No, I wasn't imagining this situation at all - I think this might be our disconnect! :)

Timelocks are pretty ephemeral things - meant to be used for one time for one vote. I don't imagine once you set this enum you will ever change its value on a given timelock and in fact I wouldn't allow it. What I would like is that different timelocks could start with different settings which may lead to variations in functionality as they move through their states.

Reply:
Gotcha, this makes a bit more sense now -- so we'd reuse the same config for different timelocks.  In that case, seems pretty neat to be able to expand that

Reply:
Yeah, the config is the global thing, that defines the "org" for the dev team. The timelocks are instances for a proposal. I'm going to be renaming all timelocks into Proposal and probably TimelockConfigs into Governances.

Reply:
I've read the MANGO governance spec and COMP, which MANGO is based on, and I think this program is actually never going to be used as is. Because it were we would have the "who will guard the guards" problem. In other words who would control upgrades of the governance program itself. It seems to me we have only two choices here:

1. Make the governance program not upgradable
2. Clone the program source and create/tweak for particular program needs

We agree that making program not upgradable is problematic but without it you would give up too much power to anybody who would control the governance program. You could for example change the governance logic and grant yourself authority to upgrade other programs.

And that leaves us with option 2), and this is, as I understand, what MANGO wants to do. They want the MANGO governance program upgrades to be controlled by the same governance program as any other program in their protocol.

Given the above I think I would go full YAGNI, and remove everything we currently don't use and which is just a placeholder for the future which might never materialise. In my opinion it would only add to a pollution to the api without no actual benefits. You can argue that it could benefit the clones but I think the clones would be even more specific.

Can we add data to an account when we upgrade program to a new version? I would assume yes because the new program would know the older accounts can have less data and apply defaults for everything missing when unpacking.
The problem I can thing about is the issue with using the account data length as identity of the account type, but with Borsh and variable data length this should be discouraged anyway, right?


Please let me know what you think,

Reply:
The people who control the upgrades of governance itself will be holders of the governance token. So governance itself is a program, whose upgradeable authority will be owned by a PDA inside Governance that is controlled by a mint, whose token holders can choose whether or not to upgrade it. Meta, right? So I think that would be Option #3 here, and I think it's fairly workable. There is also the Council mint, which is a separate mint that can also control the same PDA with a DIFFERENT pool of tokens, and the Governance program can have it's own Council made up of tokens held by important members of our community like validators and also key developers in the Mango ecosystem, which will give them great control over the governance protocol. I don't think the solution is to create for them a cloned version of governance, nor do I think they're looking for it. @bartosz-lipinski , what do you think?

Reply:
What would be the upside to MNGO token holders or other protocols to give up their full authority over the governance program and by extension over the future of the protocol itself? 
Can we prevent censorship or hostile takeover of the governance program  by unrelated or hostile 3rd parties? 
Shell we ask MNGO guys to see what they are planning to do?

Reply:
I think that's what the Council is for. The council is a very limited set of token holders and we don't mint more of them generally. Most of the chains have governance applications that are shared by more than one governance token, so this isn't a new concept here, and the social contract works. I think if governance tokens that you mention, for example, Mango or later Raydium or other projects are relying on the Validators as council holders to govern Governance , it's a pretty workable social contract. That being said, I'll have Bartosz reach out to these projects and ask them about their opinion because I do think it's totally a reasonable question to gather feedback on!